designs/2025-rule-tester-assertion-options/README.md

```diff
@@ -0,0 +1,348 @@
+   * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+   * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+   *
+   * @default false
+   */
+  requireMessage: boolean | 'message' | 'messageId';
+  /**
+   * Require full location assertions for each invalid test case.
+   * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
+   *
+   * @default false
+   */
+  requireLocation: boolean;
+} = {});
+```
+
+### Variant 2 - Test method based options
+
+```ts
+ruleTester.run("rule-name", rule, {
+  assertionOptions?: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+     * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+     * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+     *
+     * @default false
+     */
+    requireMessage?: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
+     *
+     * @default false
+     */
+    requireLocation?: boolean;
+  },
+  valid: [ /* ... */ ],
+  invalid: [ /* ... */ ],
+});
+```
+
+### Shared Logic
+
+#### requireMessage
+
+If `requireMessage` is set to `true`, the invalid test case cannot consist of an error count assertion only, but must also include a message assertion. (See below)
+This can be done either by providing only a `string`/`RegExp` message, or by using the `message`/`messageId` property of the error object in the `TestCaseError` (Same as the current behavior).
+If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+
+```ts
+ruleTester.run("rule-name", rule, {
+  assertionOptions: {
+    requireMessage: true,
+  },
+  invalid: [
+    {
+      code: "const a = 1;",
+      errors: 1, // ❌ Message isn't being checked here
+    },
+    {
+      code: "const a = 2;",
+      errors: [
+        "Error message here.", // ✅ we only check the error message
+      ],
+    },
+    {
+      code: "const a = 3;",
+      errors: [
+        {
+          message: "Error message here.", // ✅ we check the error message and potentially other properties
+        },
+      ],
+    },
+  ],
+});
+```
+
+#### requireLocation
+
+If `requireLocation` is set to `true`, the invalid test case's error validation must include all location assertions such as `line`, `column`, `endLine`, and `endColumn`.
```

Comment:
Can we clarify what this means given the discussion in https://github.com/eslint/rfcs/pull/137/files#r2276520062?

If we're saying that only `line` and `column` are required and that `endLine` and `endColumn` will only be checked if present in the error entry, let's be explicit about that.